libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Core.State
+{
+    public interface IStateStorageProvider
+    {
+        IStateStorageEntry CreateNewEntry(string stateNamespace, string key);
+
+        /// <summary>
+        /// Loads all state entries under the specified <paramref name="stateNamespace">partition</paramref>.
+        /// </summary>
+        /// <param name="stateNamespace"></param>
+        /// <returns></returns>
+        Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace);
+
+        /// <summary>
+        /// Loads a single piece of state, identified by its <paramref name="key"/> from the given <paramref name="stateNamespace">partition</paramref>.
+        /// </summary>
+        /// <param name="stateNamespace"></param>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        Task<IStateStorageEntry> Load(string stateNamespace, string key);
+
+        Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace, IEnumerable<string> keys);
+
+        Task Save(IEnumerable<IStateStorageEntry> stateStorageEntries);
+
+        Task Delete(string stateNamespace);
```

Comment:
As stated above, it's about offering the ability for optimization for bulk operation _if_ the provider can handle it, otherwise the provider emulates as it needs to... in this case maybe just calling the singular version for each one in a loop.